### PR TITLE
[PM-3797 Part 3] Add vault domains to key rotation

### PR DESCRIPTION
**Pull request by jlf0dev**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Previous PR: #3434
Adds ciphers and folders to the new key rotation.

## Code changes

* **file.ext:** Description of what was changed and why

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Infrastructure.EntityFramework/Repositories/UserRepository.cs [hunk 1]
```diff
@@ -161,6 +161,8 @@ public async Task<Core.Entities.User> GetBySsoUserAsync(string externalId, Guid?
             entity.AccountRevisionDate = user.AccountRevisionDate;
             entity.RevisionDate = user.RevisionDate;
 
+            await dbContext.SaveChangesAsync();
```
Comment (ike-kottlowski) [inline src/Infrastructure.EntityFramework/Repositories/UserRepository.cs:164]: ❓ Why is this here instead of line 180? Doesn't EF handle transactions automagically?

File: src/Api/Vault/Validators/CipherRotationValidator.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+public class CipherRotationValidator : IRotationValidator<IEnumerable<CipherWithIdRequestModel>, IEnumerable<Cipher>>
+{
+    private readonly ICipherRepository _cipherRepository;
+
+    public CipherRotationValidator(ICipherRepository cipherRepository)
+    {
+        _cipherRepository = cipherRepository;
+    }
+
+    public async Task<IEnumerable<Cipher>> ValidateAsync(User user, IEnumerable<CipherWithIdRequestModel> ciphers)
```
Comment (ike-kottlowski) [inline src/Api/Vault/Validators/CipherRotationValidator.cs:19]: ⛏️ I'm not a huge fan of having multiple return statements. 

```suggestion
public async Task<IEnumerable<Cipher>> ValidateAsync(User user, IEnumerable<CipherWithIdRequestModel> ciphers)
{
    var result = new List<Cipher>();
    if (ciphers != null || ciphers.Any())
    {
        var existingCiphers = await _cipherRepository.GetManyByUserIdAsync(user.Id);
        if (existingCiphers != null || existingCiphers.Any())
        {
            foreach (var existing in existingCiphers)
            {
                var cipher = ciphers.FirstOrDefault(c => c.Id == existing.Id);
                if (cipher == null)
                {
                    throw new BadRequestException("All existing ciphers must be included in the rotation.");
                }
                result.Add(cipher.ToCipher(existing));
            }
        }
    }

    return result;
}
```

File: src/Api/Vault/Validators/FolderRotationValidator.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+public class FolderRotationValidator : IRotationValidator<IEnumerable<FolderWithIdRequestModel>, IEnumerable<Folder>>
+{
+    private readonly IFolderRepository _folderRepository;
+
+    public FolderRotationValidator(IFolderRepository folderRepository)
+    {
+        _folderRepository = folderRepository;
+    }
+
+    public async Task<IEnumerable<Folder>> ValidateAsync(User user, IEnumerable<FolderWithIdRequestModel> folders)
```
Comment (ike-kottlowski) [inline src/Api/Vault/Validators/FolderRotationValidator.cs:19]: ⛏️ see comment for `cipherRotationValidator.cs`

**Review by ike-kottlowski — APPROVED** [review]

**Comment by jlf0dev** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3434** <a href="https://app.graphite.dev/github/pr/bitwarden/server/3434?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #3436** <a href="https://app.graphite.dev/github/pr/bitwarden/server/3436?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #3442** <a href="https://app.graphite.dev/github/pr/bitwarden/server/3442?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #3445** <a href="https://app.graphite.dev/github/pr/bitwarden/server/3445?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/bitwarden/server/3436?utm_source=stack-comment).

File: src/Api/Vault/Validators/CipherRotationValidator.cs (hunk 2 above)
Comment (jlf0dev) [reply src/Api/Vault/Validators/CipherRotationValidator.cs:19]: I'm actually a pretty big fan of early returns, which is intended to avoid more than a couple levels of nesting. I think it makes the logic harder to follow because you have to keep in mind all the previous operations.

File: src/Infrastructure.EntityFramework/Repositories/UserRepository.cs (hunk 1 above)
Comment (jlf0dev) [reply src/Infrastructure.EntityFramework/Repositories/UserRepository.cs:164]: From my [understanding](https://www.entityframeworktutorial.net/entityframework6/transaction-in-entity-framework.aspx), this will make the database call for all the previous changes to the user, separating them from all the domain calls below.
CC: @rkac-bw, let me know if I'm wrong here

Comment (rkac-bw) [reply src/Infrastructure.EntityFramework/Repositories/UserRepository.cs:164]: That is my understanding, While the overall transaction makes sure that changes are committed or rolled back as a single unit, the  SaveChangesAsync() within the try block affects when those changes are persisted to the database. Placing it within the try block allows you to make sure that changes are saved immediately instead of stored in change tracking.

Comment (rkac-bw) [reply src/Infrastructure.EntityFramework/Repositories/UserRepository.cs:164]: And it seems that the updateDataActions uses that saved data so it requires the data to be persisted

**Review by gbubemismith — APPROVED** [review]
Reviewed files owned by the vault team, and they LGTM! Nice job @jlf0dev

**Review by ike-kottlowski — APPROVED** [review]
